kvbc/src/blockchain_misc.cpp

```diff
@@ -12,10 +12,17 @@
 
 #include "blockchain_misc.hpp"
 #include "storage/merkle_tree_key_manipulator.h"
+#include "categorization/blocks.h"
 
 namespace concord::kvbc {
 
 block_version BlockVersion::getBlockVersion(const concord::kvbc::RawBlock& raw_block_ser) {
+  try {
+    const auto rawBlock = concord::kvbc::categorization::RawBlock::deserialize(raw_block_ser);
+    return concord::kvbc::block_version::V1;
+  } catch (const std::runtime_error& re) {
+  }
+
```

Comment:
regarding ```block_version BlockVersion::getBlockVersion(const char* raw_block_ser, size_t len) {
  ConcordAssertGE(len, BLOCK_VERSION_SIZE);
  return *(reinterpret_cast<const block_version*>(raw_block_ser));
}```

is it a problem that it can return values that are not part of the enum?

Reply:
This is a very good que. To check this i used the following prog to understand:

`#include<ios>
#include<iostream>
#include<cstdlib>

using namespace std;

using version_type = uint16_t;
enum class block_version : version_type { V1 = 0x0, V4 = 0x1, V5 = 0x2, V6 = 0x3 };
const size_t BLOCK_VERSION_SIZE = sizeof(version_type);

block_version getBlockVersion(const char* raw_block_ser, size_t len) {
  return *(reinterpret_cast<const block_version*>(raw_block_ser));
}

void printBlockVer(block_version bv) {
        switch(bv) {
                case block_version::V1 :
                        cout<<"V1"<<endl;
                        break;
                case block_version::V4 :
                        cout<<"V4"<<endl;
                        break;
                case block_version::V5 :
                        cout<<"V5"<<endl;
                        break;
                case block_version::V6 :
                        cout<<"V6"<<endl;
                        break;
                default:
                        cout<<"Others: "<<hex<<*(reinterpret_cast<uint16_t *>(&bv))<<endl;
        }
}

int main(void) {
        cout<<"Size = "<<BLOCK_VERSION_SIZE<<endl;

        uint16_t c1=0x0;
        uint16_t c2=0x1;
        uint16_t c3=0x2;
        uint16_t c4=0x3;
        uint16_t c5=0x4;
        uint16_t c6=0x5;
        uint16_t c7=0x10;
        uint16_t c8=0x23;

        printBlockVer(getBlockVersion(reinterpret_cast<const char*>(&c1), 2));
        printBlockVer(getBlockVersion(reinterpret_cast<const char*>(&c2), 2));
        printBlockVer(getBlockVersion(reinterpret_cast<const char*>(&c3), 2));
        printBlockVer(getBlockVersion(reinterpret_cast<const char*>(&c4), 2));
        printBlockVer(getBlockVersion(reinterpret_cast<const char*>(&c5), 2));
        printBlockVer(getBlockVersion(reinterpret_cast<const char*>(&c6), 2));
        printBlockVer(getBlockVersion(reinterpret_cast<const char*>(&c7), 2));
        printBlockVer(getBlockVersion(reinterpret_cast<const char*>(&c8), 2));

        return 0;
}`

Below is the o/p
achaudhuri@ubuntu:~/data/prog/enum_check$ ./enum_check 
Size = 2
V1
V4
V5
V6
Others: 4
Others: 5
Others: 10
Others: 23

So we can see that the outside enums are not converted to the correct enum and thus this should be fine.

Reply:
thanks @cloudnoize I have updated the change accordingly.